File: csharp/1049.last-stone-weight-ii.cs
/*
 * @lc app=leetcode id=1049 lang=csharp
 *
 * [1049] Last Stone Weight II
 */

using System.Linq;

// @lc code=start
public class Solution {
    public int LastStoneWeightII(int[] stones) {
        int sum = stones.Sum(), h = sum / 2;
        bool[] dp = new bool[h + 1];
        dp[0] = true;
        foreach (int stone in stones) {
            for (int s = h; s >= stone; --s) {
                dp[s] = dp[s] || dp[s - stone];
            }
        }
        for (int s = h; s >= 0; --s) {
            if (dp[s]) {
                return sum - s * 2;
            }
        }
        return -1;
    }
}
// @lc code=end

File: java/1049.last-stone-weight-ii.java
/*
 * @lc app=leetcode id=1049 lang=java
 *
 * [1049] Last Stone Weight II
 */

// @lc code=start
class Solution {
    public int lastStoneWeightII(int[] stones) {
        int sum = 0;
        for (int s : stones) {
            sum += s;
        }
        int half = sum / 2;
        boolean[] dp = new boolean[half + 1];
        dp[0] = true;
        for (int s : stones) {
            for (int t = half; t >= s; --t) {
                dp[t] |= dp[t - s];
            }
        }
        for (int t = half; t >= 0; --t) {
            if (dp[t]) {
                return sum - 2 * t;
            }
        }
        throw new IllegalStateException();
    }
}
// @lc code=end

